vj4/ui/components/ide/ide.js

```diff
@@ -0,0 +1,418 @@
+import React from 'react';
+import './ide.styl';
+import Codemirror from 'react-codemirror';
+import RecordRow from './record-row';
+import _ from 'lodash';
+import sockJS from 'sockjs-client';
+import { post } from '../../misc/Util';
+import Constants from './ide-constants';
+
+import 'codemirror/mode/clike/clike';
+import 'codemirror/mode/pascal/pascal';
+import 'codemirror/mode/python/python';
+import 'codemirror/mode/javascript/javascript';
+import 'codemirror/mode/ruby/ruby';
+import 'codemirror/mode/php/php';
+
+class Ide extends React.Component {
+
+  static spanSupportLanguageList() {
+    return Object.keys(Constants.supportLangTypes).map((key, idx) =>
+      <option value={key} key={idx}>{Constants.supportLangTypes[key]}</option>
+    );
+  }
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      displayTesting: true,
+      displayEvalutating: true,
+      code: '',
+      currentTestingInputText: '',
+      currentTestingOutputText: '',
+      language: 'c',
+      testingData: [],
+      resultData: [],
+      currentActiveTabIndex: -1,
+      codeAreaOptions: {
+        lineNumbers: true,
+        matchBrackets: true,
+        readOnly: false,
+        indentUnit: 4,
+        // mode: Ide.mapLangToMode[this.state.language],
+        extraKeys: {
+          Tab: (cm) => {
+            const spaces = Array(cm.getOption('indentUnit') + 1).join(' ');
+            cm.replaceSelection(spaces, 'end', '+input');
+          },
+        },
+      },
+    };
+    this.sock = null;
+  }
+
+  componentDidMount() {
+    if (!this.sock) {
+      this.sock = sockJS('/records-conn');
+      this.sock.onmessage = (message) => {
+        const rdoc = JSON.parse(message.data).rdoc;
+        // console.log('Received rdoc = ', rdoc);
+        /* Check uid and pid to ensure that the record should be shown */
+        if ((window.UserContext.uid.toString() !== rdoc.udoc._id.toString())
+          || (window.Context.problemId.toString() !== rdoc.pid.toString())) {
+          return;
+        }
+        const nextResultData = this.state.resultData.slice();
+        const existedRdoc = nextResultData.filter(val => val._id === rdoc._id)[0];
+        if (existedRdoc) {
+          const idx = nextResultData.indexOf(existedRdoc);
+          nextResultData[idx] = rdoc;
+        } else {
+          nextResultData.unshift(rdoc);
+        }
+        return this.setState({
+          resultData: nextResultData,
+        });
+      };
+    }
+  }
+
+  componentWillUnmount() {
+    if (this.sock) {
+      this.sock.close();
+      this.sock = null;
+    }
+  }
+
+  onLeaveIDEMode() {
+    if (window.confirm('确认退出IDE模式吗？')) {
+      $('body').removeClass('ide-mode');
```

Comment:
In fact, you are re-implementing `leaveIdeMode`. Since the next line triggers a `hashchange`, you'd better call `leaveIdeMode` in `hashchange`